flink-state-backends/flink-statebackend-forst/src/main/java/org/apache/flink/state/forst/ForStKeyedStateBackend.java

```diff
@@ -372,20 +379,32 @@ public <N, S extends InternalKeyedState, SV> S createStateInternal(
                             Tuple2.of(oldStateInfo.columnFamilyHandle, castedMetaInfo),
                             stateDesc,
                             stateSerializer,
+                            userKeySerializer,
                             namespaceSerializer);
 
             newStateInfo =
                     new ForStOperationUtils.ForStKvStateInfo(
                             oldStateInfo.columnFamilyHandle, newMetaInfo);
             kvStateInformation.put(stateDesc.getStateId(), newStateInfo);
         } else {
-            newMetaInfo =
-                    new RegisteredKeyValueStateBackendMetaInfo<>(
-                            stateDesc.getStateId(),
-                            stateDesc.getType(),
-                            namespaceSerializer,
-                            stateSerializer,
-                            StateSnapshotTransformer.StateSnapshotTransformFactory.noTransform());
+            if (stateDesc.getType().equals(StateDescriptor.Type.MAP)) {
+                newMetaInfo =
+                        new RegisteredKeyAndUserKeyValueStateBackendMetaInfo<>(
+                                stateDesc.getStateId(),
+                                stateDesc.getType(),
+                                namespaceSerializer,
+                                stateSerializer,
+                                userKeySerializer);
+            } else {
+                newMetaInfo =
+                        new RegisteredKeyValueStateBackendMetaInfo<>(
+                                stateDesc.getStateId(),
+                                stateDesc.getType(),
+                                namespaceSerializer,
+                                stateSerializer,
+                                StateSnapshotTransformer.StateSnapshotTransformFactory
+                                        .noTransform());
+            }
```

Comment:
Could we align these two constructor? Either both provide `StateSnapshotTransformer` or both not